glance_store/_drivers/s3.py

```diff
@@ -551,63 +626,71 @@ def add_singlepart(self, image_file, bucket_obj, obj_name, loc, verifier):
         # take this opportunity to calculate the image checksum while
         # writing the tempfile, so we don't need to call key.compute_md5()
 
-        msg = ("Writing request body file to temporary file "
-               "for %s") % self._sanitize(loc.get_uri())
+        msg = "Writing request body file to temporary file for %s" % self._sanitize(loc.get_uri())
         LOG.debug(msg)
 
         tmpdir = self.s3_store_object_buffer_dir
         temp_file = tempfile.NamedTemporaryFile(dir=tmpdir)
+        os_hash_value = hashlib.new(str(hashing_algo))
         checksum = hashlib.md5()
         for chunk in utils.chunkreadable(image_file, self.WRITE_CHUNKSIZE):
+            os_hash_value.update(chunk)
             checksum.update(chunk)
             if verifier:
                 verifier.update(chunk)
             temp_file.write(chunk)
         temp_file.flush()
 
-        msg = ("Uploading temporary file to S3 "
-               "for %s") % self._sanitize(loc.get_uri())
+        msg = "Uploading temporary file to S3 for %s" % self._sanitize(loc.get_uri())
         LOG.debug(msg)
 
         # OK, now upload the data into the key
-        key.set_contents_from_file(open(temp_file.name, 'rb'),
-                                   replace=False)
-        size = key.size
+        s3_client.put_object(Body=open(temp_file.name, 'rb'), Bucket=bucket_id, Key=object_key)
+        key = get_key(s3_client, bucket_id, object_key)
+        size = key['ContentLength']
+        hash_hex = os_hash_value.hexdigest()
         checksum_hex = checksum.hexdigest()
 
-        LOG.debug("Wrote %(size)d bytes to S3 key named %(obj_name)s "
-                  "with checksum %(checksum_hex)s" %
-                  {'size': size,
-                   'obj_name': obj_name,
-                   'checksum_hex': checksum_hex})
+        metadata = {}
+        if self.backend_group:
+            metadata['backend'] = u"%s" % self.backend_group
+
+        LOG.debug("""
+                  Wrote %(size)d bytes to S3 key named %(object_key)s
+                  with checksum %(checksum_hex)s
+                  """ % {'size': size,
+                         'object_key': object_key,
+                         'checksum_hex': checksum_hex})
 
-        return (loc.get_uri(), size, checksum_hex, {})
+        return (loc.get_uri(), size, checksum_hex, hash_hex, metadata)
```

Comment:
タプルの丸括弧は冗長だと思います。他にも同様の箇所があるので確認をお願いします。
```suggestion
        return loc.get_uri(), size, checksum_hex, hash_hex, metadata
```